File: leetcode/73-set-matrix-zeroes.cpp
/*
Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in place.
*/

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        if (matrix.size() == 0) {
            return;
        }
        vector<bool> rows(matrix.size(), false);
        vector<bool> cols(matrix[0].size(), false);

        for (int i=0; i<matrix.size(); ++i) {
            for (int j=0; j<matrix[0].size(); ++j) {
                if (matrix[i][j] == 0) {
                    rows[i] = true;
                    cols[j] = true;
                }
            }
        }
        for (int i=0; i<rows.size(); ++i) {
            if (rows[i]) {
                fill(matrix[i].begin(), matrix[i].end(), 0);
            }
        }
        for (int j=0; j<cols.size(); ++j) {
            if (cols[j]) {
                for (int i=0; i<matrix.size(); ++i) {
                    matrix[i][j] = 0;
                }
            }
        }
    }
};

File: leetcode/73-set-matrix-zeroes.rb
# @param {Integer[][]} matrix
# @return {Void} Do not return anything, modify matrix in-place instead.
def set_zeroes(matrix)
  m = matrix.size
  n = matrix[0].size
  (0...m).each do |i|
    (0...n).each do |j|
      set_nil(matrix, i, j, m, n) if matrix[i][j] == 0
    end
  end
  (0...m).each do |i|
    (0...n).each do |j|
      matrix[i][j] = 0 if matrix[i][j].nil?
    end
  end
end


def set_nil(matrix, i, j, m, n)
  (0...m).each do |x|
    matrix[x][j] = nil if matrix[x][j] != 0
  end
  (0...n).each do |x|
    matrix[i][x] = nil if matrix[i][x] != 0
  end
end
